dig_test.go

```diff
@@ -0,0 +1,360 @@
+			return map[string]string{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(m map[string]string) {
+			require.NotNil(t, m, "invoke got zero value map")
+		}), "invoke failed")
+	})
+
+	t.Run("channel", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		ch := make(chan int)
+		require.NoError(t, c.Provide(ch), "provide failed")
+		require.NoError(t, c.Invoke(func(ch chan int) {
+			require.NotNil(t, ch, "invoke got nil chan")
+		}), "invoke failed")
+	})
+
+	t.Run("channel constructor", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		c := New()
+		require.NoError(t, c.Provide(func() chan int {
+			return make(chan int)
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(ch chan int) {
+			require.NotNil(t, ch, "invoke got nil chan")
+		}), "invoke failed")
+	})
+
+	t.Run("func constructor", func(t *testing.T) {
+		t.Skip("Not yet supported.")
+		// Functions passed directly to Provide are treated as constructors,
+		// but we can still put functions into the container with constructors.
+		// This makes injecting builders simple.
+		c := New()
+		require.NoError(t, c.Provide(func() func(int) {
+			return func(int) {}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(f func(int)) {
+			require.NotNil(t, f, "invoke got nil function pointer")
+		}), "invoke failed")
+	})
+
+	t.Run("interface", func(t *testing.T) {
+		// TODO: This doesn't work as a reasonable user would expect, since
+		// passing an io.Writer as interface{} erases information. (Put
+		// differently, we go from having an io.Writer satisfied by a
+		// *bytes.Buffer to having an interface{} satisfied by *bytes.Buffer;
+		// the fact that the io.Writer interface was involved is lost forever.)
+		c := New()
+		var w io.Writer = &bytes.Buffer{}
+		require.NoError(t, c.Provide(w), "provide failed")
+		require.Error(t, c.Invoke(func(w io.Writer) {
+		}), "expected relying on provided interface to fail")
+		require.NoError(t, c.Invoke(func(b *bytes.Buffer) {
+			assert.NotNil(t, b, "expected to get concrete type of interface")
+		}), "expected relying on concrete type to succeed")
+
+	})
+
+	t.Run("interface constructor", func(t *testing.T) {
+		c := New()
+		require.NoError(t, c.Provide(func() io.Writer {
+			return &bytes.Buffer{}
+		}), "provide failed")
+		require.NoError(t, c.Invoke(func(w io.Writer) {
+			require.NotNil(t, w, "invoke got nil interface")
+		}), "invoke failed")
+	})
+
+	t.Run("multiple-type constructor", func(t *testing.T) {
+		c := New()
+		constructor := func() (*bytes.Buffer, []int, error) {
+			return &bytes.Buffer{}, []int{42}, nil
+		}
+		consumer := func(b *bytes.Buffer, nums []int) {
+			assert.NotNil(t, b, "invoke got nil buffer")
+			assert.Equal(t, 1, len(nums), "invoke got empty slice")
+		}
+		require.NoError(t, c.Provide(constructor), "provide failed")
+		require.NoError(t, c.Invoke(consumer), "invoke failed")
+	})
+
+	t.Run("collections and instances of same type", func(t *testing.T) {
```

Comment:
The test points out that dig's pointer and map support is currently a little broken - it's not possible to provide both `[]*Foo` and `*Foo`. Obviously, we *should* support providing both.